### PR TITLE
Improved handling of failed cases in recurr.rsolve_hyper

### PR DESCRIPTION
**Pull request by leosartaj**
Previously, following cases raised error, now they return `None`

```
In [1]: r1 = (k + 1)*g(k)

In [2]: rsolve(r1, g(k))
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-2-9a0672b637aa> in <module>()
----> 1 rsolve(r, g(k))

/home/leosartaj/projects/sympy/sympy/solvers/recurr.pyc in rsolve(f, y, init)
    785     coeffs = [H_part[i] for i in range(K_max + 1)]
    786 
--> 787     result = rsolve_hyper(coeffs, -i_part, n, symbols=True)
    788 
    789     if result is None:

/home/leosartaj/projects/sympy/sympy/solvers/recurr.pyc in rsolve_hyper(coeffs, f, n, **hints)
    606                 degrees.append(polys[i].degree())
    607 
--> 608         d, poly = max(degrees), S.Zero
    609 
    610         for i in range(0, r + 1):

ValueError: max() arg is an empty sequence
```

```
In [3]: r2 = (k + 1)*g(k) + (k + 3)*g(k + 1) + (k + 5)*g(k + 2)

In [4]: rsolve(r2, g(k))
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
<ipython-input-4-20f69c2be462> in <module>()
----> 1 rsolve(r2, g(k))

/home/leosartaj/projects/sympy/sympy/solvers/recurr.pyc in rsolve(f, y, init)
    785     coeffs = [H_part[i] for i in range(K_max + 1)]
    786 
--> 787     result = rsolve_hyper(coeffs, -i_part, n, symbols=True)
    788 
    789     if result is None:

/home/leosartaj/projects/sympy/sympy/solvers/recurr.pyc in rsolve_hyper(coeffs, f, n, **hints)
    630                 n0 = 0
    631                 for n_root in roots(ratio.as_numer_denom()[1], n).keys():
--> 632                     if (n0 < (n_root + 1)) == True:
    633                         n0 = n_root + 1
    634                 K = product(ratio, (n, n0, n - 1))

/home/leosartaj/projects/sympy/sympy/core/expr.pyc in __gt__(self, other)
    278             if (me.is_complex and me.is_real is False) or \
    279                     me.has(S.ComplexInfinity):
--> 280                 raise TypeError("Invalid comparison of complex %s" % me)
    281             if me is S.NaN:
    282                 raise TypeError("Invalid NaN comparison")

TypeError: Invalid comparison of complex -1 - sqrt(3)*I/3

```

Now, 

```
>>> rsolve(r1, g(k))
None
>>> rsolve(r2, g(k))
None
```

Ping @flacjacket @jcrist

### CODE REVIEW

File: sympy/solvers/recurr.py [hunk 1]
```diff
@@ -605,7 +605,10 @@ def rsolve_hyper(coeffs, f, n, **hints):
             if not poly.is_zero:
                 degrees.append(polys[i].degree())
 
-        d, poly = max(degrees), S.Zero
+        if degrees:
+            d, poly = max(degrees), S.Zero
+        else:
+            return None
```
Comment (aktech) [inline sympy/solvers/recurr.py]: Does this mean, solution doesn't exist, or we are not able to find it ?

Comment (leosartaj) [inline sympy/solvers/recurr.py]: > Does this mean, solution doesn't exist, or we are not able to find it ?

We are not able to find it currently.

File: sympy/solvers/tests/test_recurr.py [hunk 2]
```diff
@@ -184,6 +184,10 @@ def test_rsolve():
             (C0*((sqrt(-a**2 + 1) - 1)/a)**n +
              C1*((-sqrt(-a**2 + 1) - 1)/a)**n)).simplify() == 0
 
+    assert rsolve((k + 1)*y(k), y(k)) is None
+    assert (rsolve((k + 1)*y(k) + (k + 3)*y(k + 1) + (k + 5)*y(k + 2), y(k))
+            is None)
```
Comment (aktech) [inline sympy/solvers/tests/test_recurr.py]: So, I guess here also the test says we are not able to find the solution, so we are returning `None`.
I would rather `XFAIL` these in a separate function, so that it may eventually pass someday, when `recurr.py` module is improved. Is there anything stopping us from doing this?

Comment (leosartaj) [inline sympy/solvers/tests/test_recurr.py]: Well, I did this due to the failures in computing ``FormalPowerSeries``. The algorithm converts a function into a RE eventually. That I try to solve using a function. If it's unsuccessful I fallback to more general ``rsolve``. 

I really don't like the failures.

The first RE is what you get when the function is ``1/x``. I will like it to fail silently.
In this particular case, I can check if there is a single term, don't compute.

The second RE is computed when the function is ``(x**2 + x + 1)/x**3``. Now this kind of case is hard to capture. As, many RE of this form are solved RE.

One way is to return ``None`` like I did. Otherwise we can also raise an explicit exception.  In general, ``rsolve`` returns ``None`` whenever it's not able to compute a valid RE.

What do you think?

Comment (aktech) [inline sympy/solvers/tests/test_recurr.py]: > As, many RE of this form are solved RE.

what does this mean?

> In general, rsolve returns None whenever it's not able to compute a valid RE.

Yes, It didn't have a single case raising an exception other than `ValueError` , though I don't understand why is it so, One possibility is that It possibly handles a particular class of problems very well, so we don't see much failures in that domain, so no exception in that case except for some newly explored cases like these.

Your fix looks consistent with the current code, but in general silencing an exception has its own consequences, so I am not sure if it's right to silent one here.

Comment (leosartaj) [inline sympy/solvers/tests/test_recurr.py]: > As, many RE of this form are solved RE.
> what does this mean?

Sorry, I meant As, many RE of this form are solved by ``rsolve``.

> In general, rsolve returns None whenever it's not able to compute a valid RE.
Yes, It didn't have a single case raising an exception other than ValueError , though I don't understand why is it so, One possibility is that It possibly handles a particular class of problems very well, so we don't see much failures in that domain, so no exception in that case except for some newly explored cases like these.

Well, SymPy does this a lot. Whenever it's not able to compute say an ``integral``. It returns the expr as it is.

> Your fix looks consistent with the current code, but in general silencing an exception has its own consequences, so I am not sure if it's right to silent one here.

Well, we can raise an exception. LIke ``ValueError("Unable to compute degree")``?

Comment (leosartaj) [inline sympy/solvers/tests/test_recurr.py]: But I have seen. SymPy prefers returning the ``expr`` as it is, then raising an exception if it's unable to compute something and lacks an algorithm.

Comment (aktech) [inline sympy/solvers/tests/test_recurr.py]: > Well, SymPy does this a lot. Whenever it's not able to compute say an integral. It returns the expr as it is.

That's basically an unevaluated object it returns, when it doesn't knows how to further process (evaluate) the input.

say for e.g.
```python
def foo(user_input = i):
    if i is this1:
        do this1
    if i is this2:
        do this2:
            while doing this2:
                if this_x fails: raise NotImplementedError
    if i is this3:
        do this
    else
        return Foo(i)  # unevaluated object
```

Otherwise it raises `ValueError` if input is invalid, as per my understanding.

> Well, we can raise an exception. LIke ValueError("Unable to compute degree")?

Maybe, but since I am unaware of the algorithm, so I am not sure if it's the right candidate.

Comment (leosartaj) [inline sympy/solvers/tests/test_recurr.py]: Docstring of recurr module

> This module is intended for solving recurrences or, in other words,
    difference equations. Currently supported are linear, inhomogeneous
    equations with polynomial or rational coefficients.

I guess they fall in this category. So, input is valid.

> Maybe, but since I am unaware of the algorithm, so I am not sure if it's the right candidate.

Well, somebody else can help out then.

Thanks btw!

Comment (aktech) [inline sympy/solvers/tests/test_recurr.py]: > I guess they fall in this category. So, input is valid.

In that case `NotImplementedError` looks like a good candidate.

> Well, somebody else can help out then.

Yeah, Indeed.
Thanks for addressing my concerns.

**Comment by leosartaj** [comment]
Ping @jcrist @flacjacket

**Comment by jcrist** [comment]
The fix looks solid. I can see @aktech's point though - it appears we're testing that the solver fails to find a solution - which is a bad test. However, it also seems that to be able to solve these problems a modest rework of the solver will be needed, so many tests will need to be rewritten anyway. What you have now may not actually be bad, but a separate xfailed test containing the tests for `None` certainly couldn't hurt.

**Comment by leosartaj** [comment]
To be clear I should mark the added tests as xfailed and check them against ``None`` right? They will pass though but I guess it will be better to keep them as xfailed.

**Comment by jcrist** [comment]
Oh, I guess I didn't think that through clearly, it's kind of late here :). The tests are fine as they are I suppose. Merging.

**Comment by leosartaj** [comment]
Thanks!
